lib/datasource.js

```diff
@@ -2115,26 +2157,116 @@ DataSource.prototype.transaction = function() {
 
   transaction.isTransaction = true;
   transaction.origin = dataSource;
-  transaction.name = dataSource.name;
-  transaction.settings = dataSource.settings;
   transaction.connected = false;
   transaction.connecting = false;
-  transaction.connector = dataSource.connector.transaction();
 
   // create blank models pool
-  transaction.modelBuilder = new ModelBuilder();
-  transaction.models = transaction.modelBuilder.models;
-  transaction.definitions = transaction.modelBuilder.definitions;
+  var modelBuilder = new ModelBuilder();
+  var slaveModels = modelBuilder.models;
+  transaction.modelBuilder = modelBuilder;
+  transaction.models = slaveModels;
+  transaction.definitions = modelBuilder.definitions;
+
+  // For performance reasons, use a getter per model and only call copyModel()
+  // for the models that are actually used. These getters are then replaced
+  // with the actual values on first use.
+  var masterModels = dataSource.modelBuilder.models;
+  Object.keys(masterModels).forEach(function(name) {
+    Object.defineProperty(slaveModels, name, {
+      enumerable: true,
+      configurable: true,
+      get: function() {
+        // Delete getter so copyModel() can redefine slaveModels[name].
+        delete slaveModels[name];
+        return dataSource.copyModel.call(transaction, masterModels[name]);
```

Comment:
I find this suspicious - we delete `slaveModels[name]` but don't set any new value. Should we store the result of `dataSource.copyModel` to `slaveModels[name]`?

Reply:
`dataSource.copyModel()` sets the new value by itself. I need to delete it so that it can do that, because if it's still there, it wouldn't override. That's part of the caching mechanism, and works well.

Reply:
I could store it additionally for clarity, but it's not required...

Reply:
Sounds good. Is there any test verifying this? (I.e. accessing the same slave model multiple times?)

Reply:
That's a good point! I will add a test for it.